### PR TITLE
add scanvi to scib

### PR DESCRIPTION
**Pull request by danielStrobl**
separate PR for scanVI

### CODE REVIEW

**Comment by LuckyMD** [comment]
@romain-lopez it would be great if you could take a look at the function here to see if this is the parametrization you would expect for SCANVI.

File: scIB/integration.py [hunk 1]
```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
...
+    del net_adata.raw # Note that this only works from anndata 0.7
+
+    # Define batch indices
+    le = LabelEncoder()
+    net_adata.obs['batch_indices'] = le.fit_transform(net_adata.obs[batch].values)
+    net_adata.obs['labels'] = le.fit_transform(net_adata.obs[labels].values)
+
+    net_adata_out = AnnDatasetFromAnnData(net_adata)
+    net_adata_out.labels = net_adata.obs['labels']
+
```
Comment (romain-lopez) [inline scIB/integration.py:211]: Usually, we learn scVI first, and then on the initialized weights of scVI run the scANVI training. I need Pierre to verify this.

File: scIB/integration.py [hunk 2]
```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
...
+        raise TypeError('Adata does not contain a `counts` layer in `adata.layers[`counts`]`')
+
+    from scvi.models import VAE, SCANVI
+    from scvi.inference import AlternateSemiSupervisedTrainer, SemiSupervisedTrainer
+    from sklearn.preprocessing import LabelEncoder
+    from scvi.dataset import AnnDatasetFromAnnData
+
+    # Defaults from SCVI github tutorials scanpy_pbmc3k and harmonization
+    n_epochs=np.min([round((20000/adata.n_obs)*400), 400])
+    n_latent=30
```
Comment (romain-lopez) [inline scIB/integration.py:194]: n_latent are defined here but not fed into the model. I guess it depends what you're evaluating.

File: scIB/integration.py [hunk 3]
```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
...
+    net_adata_out = AnnDatasetFromAnnData(net_adata)
+    net_adata_out.labels = net_adata.obs['labels']
+
+    scanvi = SCANVI(
+        net_adata_out.nb_genes,
+        reconstruction_loss='nb',
+        n_batch=net_adata_out.n_batches,
+        n_labels=net_adata_out.n_labels,
+        n_layers=n_layers,
+        #n_latent=n_latent,
```
Comment (romain-lopez) [inline scIB/integration.py:218]: those commented lines seem dangerous, see my other comment

File: scIB/integration.py [hunk 4]
```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
...
+        #n_hidden=n_hidden,
+    )
+
+    trainer = SemiSupervisedTrainer(
+        scanvi,
+        net_adata_out,
+        train_size=1.0,
+        n_epochs_classifier=10,
```
Comment (romain-lopez) [inline scIB/integration.py:226]: there are many trainers and parameters. I need Pierre to confirm which one to use (I am not sure myself)

File: scIB/integration.py [hunk 5]
```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
...
+        use_cuda=False,
+    )
+    
+    trainer.labelled_set = trainer.create_posterior()
```
Comment (romain-lopez) [inline scIB/integration.py:230]: That seems definitely super suspect. In scANVI you need to redefine what is the labelled set and what is the unlabelled set systematically. I don't know what this line would do exactly. Does it just use all cells as labelled? 

Which cell type information do you want to give to scANVI? All the cell labels? Or only a few of them? Do you have some "unlabelled" inside of the labels field?

Usually we use only label from one dataset for example or only a couple of labels from a single dataset. It does not seem meaningful to use scANVI like this? I'm unsure. Let me know if what I say makes sense.

File: scIB/integration.py [hunk 6]
```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
...
+
+    trainer.train(n_epochs=10)
```
Comment (romain-lopez) [inline scIB/integration.py:232]: same here, n_epochs is defined but is not used. Also, I'll need Pierre here to confirm how to adjust this number.

File: scIB/integration.py (hunk 5 above)
Comment (danielStrobl) [reply scIB/integration.py:230]: Hey! Thanks for your comments!

We would want to all of the labels that we have for integration, we would not have any unlabelled cells. What we basically wanted to to is to use scANVI in a similar way to scVI but with added cell type labels.
I'm not really sure if we can use scANVI in such a way.

File: scIB/integration.py (hunk 1 above)
Comment (LuckyMD) [reply scIB/integration.py:211]: I see this in the tutorial now, yes... this should be added. Thanks for pointing it out. I guess that makes both methods intertwined (as they are anyway), but that's fine if this is the canonical way to run scANVI.

File: scIB/integration.py (hunk 5 above)
Comment (LuckyMD) [reply scIB/integration.py:230]: The though process here is that adding more prior information, on which you condition your latent variable z will improve the latent representation. So we pass all labels to get the best possible representation. If this approach doesn't make sense to you, maybe we can also omit scANVI and only compare scVI though, as then conceptually there might not be much difference except the added layer of complexity.

Comment (romain-lopez) [reply scIB/integration.py:230]: Hi Malte, 

I am going to add in this PR a piece of code to do that. The idea is that the labels should help structure the latent space better, and be some form of additional information. 

One thing that would be good to add is that usually, scANVI is used for probabilistic annotations of cells, and when it is fed the complete annotation, it won't be probabilistic in the cell types anymore. Consequently, the latent space becomes biased with respect to the fed labels. Hopefully, this model should get better clustering results than scVI for example. It might deform the latent space in other ways also. 

Also, I would insist on the fact that usually not all the labels are given. But I agree this is nice to add it as a "cell type informed" latent space and see what comes out of it.

File: scIB/integration.py (hunk 1 above)
Comment (romain-lopez) [reply scIB/integration.py:211]: yes, we have been experimenting with training procedures for quite a while. This is the most stable way of running the method.

**Comment by romain-lopez** [comment]
Hi, please find a code snippet below, on how to run scanvi properly in this setting
```

def runScanvi(adata, batch, labels, hvg=True):
    # Use non-normalized (count) data for scanvi!
    
    # Check for counts data layer
    if 'counts' not in adata.layers:
        raise TypeError('Adata does not contain a `counts` layer in `adata.layers[`counts`]`')

    from scvi.models import VAE, SCANVI
    from scvi.inference import UnsupervisedTrainer, SemiSupervisedTrainer
    from sklearn.preprocessing import LabelEncoder
    from scvi.dataset import AnnDatasetFromAnnData
    
    # STEP 1: prepare the data
    net_adata = adata.copy()
    net_adata.X = adata.layers['counts']
    del net_adata.layers['counts']
    # Ensure that the raw counts are not accidentally used
    del net_adata.raw # Note that this only works from anndata 0.7

    # Define batch indices
    le = LabelEncoder()
    net_adata.obs['batch_indices'] = le.fit_transform(net_adata.obs[batch].values)
    net_adata.obs['labels'] = le.fit_transform(net_adata.obs[labels].values)

    net_adata = AnnDatasetFromAnnData(net_adata)

    print("scANVI dataset object with {} batches and {} cell types".format(net_adata.n_batches, net_adata.n_labels))

    if hvg is True:
        # this also corrects for different batches by default
        net_adata.subsample_genes(2000, mode="seurat_v3")

    # # Defaults from SCVI github tutorials scanpy_pbmc3k and harmonization
    n_epochs_scVI = np.min([round((20000/adata.n_obs)*400), 400])
    n_epochs_scANVI = np.min([10, np.max([2, round(n_epochs_scVI / 3.)])])
    n_latent=30
    n_hidden=128
    n_layers=2


    # STEP 2: RUN scVI to initialize scANVI

    vae = VAE(
        net_adata.nb_genes,
        reconstruction_loss='nb',
        n_batch=net_adata.n_batches,
        n_latent=n_latent,
        n_hidden=n_hidden,
        n_layers=n_layers,
    )

    trainer = UnsupervisedTrainer(
        vae,
        net_adata,
        train_size=1.0,
        use_cuda=True,
    )

    trainer.train(n_epochs=n_epochs_scVI, lr=1e-3)

    # STEP 3: RUN scANVI

    scanvi = SCANVI(net_adata.nb_genes, net_adata.n_batches, net_adata.n_labels,
                          n_hidden=n_hidden, n_latent=n_latent, n_layers=n_layers, dispersion='gene')
    scanvi.load_state_dict(trainer.model.state_dict(), strict=False)

    # use default parameter from semi-supervised trainer class
    trainer_scanvi = SemiSupervisedTrainer(scanvi, net_adata)
    # use all cells as labelled set
    trainer_scanvi.labelled_set = trainer_scanvi.create_posterior(trainer_scanvi.model, net_adata, indices=np.arange(len(net_adata)))
    # put one cell in the unlabelled set
    trainer_scanvi.unlabelled_set = trainer_scanvi.create_posterior(indices=[0])
    trainer_scanvi.train(n_epochs=n_epochs_scANVI)

    # extract info from posterior
    scanvi_full = trainer_scanvi.create_posterior(trainer_scanvi.model, net_adata, indices=np.arange(len(net_adata)))
    latent, _, _ = scanvi_full.sequential().get_latent()

    adata.obsm['X_emb'] = latent

    return adata, scanvi_full.sequential(), trainer_scanvi
```

**Comment by romain-lopez** [comment]
Ah, I would check this carefully, I have changed a few things while testing inadvertantly like "hvg=True"

**Comment by danielStrobl** [comment]
tested and it's working fine

File: scIB/integration.py [hunk 7]
```diff
@@ -192,7 +192,88 @@ def runScvi(adata, batch, hvg=None):
     adata.obsm['X_emb'] = latent
 
     return adata
+def runScanvi(adata, batch, labels, hvg=None):
```
Comment (LuckyMD) [inline scIB/integration.py:195]: `hvg` is not used in the function. Could remove this argument.

File: scIB/integration.py [hunk 8]
```diff
@@ -192,7 +192,88 @@ def runScvi(adata, batch, hvg=None):
...
+
+    print("scANVI dataset object with {} batches and {} cell types".format(net_adata.n_batches, net_adata.n_labels))
+
+    #if hvg is True:
+    #    # this also corrects for different batches by default
+    #    net_adata.subsample_genes(2000, mode="seurat_v3")
+
+    # # Defaults from SCVI github tutorials scanpy_pbmc3k and harmonization
+    n_epochs_scVI = np.min([round((20000/adata.n_obs)*400), 400]) #400
+    n_epochs_scANVI = np.min([10, np.max([2, round(n_epochs_scVI / 3.)])])
```
Comment (LuckyMD) [inline scIB/integration.py:230]: `n_epochs_scANVI` seems very low as a default. This is definitely okay @romain-lopez?

File: scIB/integration.py [hunk 9]
```diff
@@ -192,7 +192,88 @@ def runScvi(adata, batch, hvg=None):
...
+        n_latent=n_latent,
+        n_hidden=n_hidden,
+        n_layers=n_layers,
+    )
+
+    trainer = UnsupervisedTrainer(
+        vae,
+        net_adata,
+        train_size=1.0,
+        use_cuda=True,
```
Comment (LuckyMD) [inline scIB/integration.py:251]: `use_cuda` sounds like GPU, which we don't have in our setup on the server. Does it still work without GPU available? We have `use_cude=False` for scVI defaults in the `runScvI()` function.

File: scIB/integration.py [hunk 10]
```diff
@@ -192,7 +192,88 @@ def runScvi(adata, batch, hvg=None):
...
+    trainer_scanvi.train(n_epochs=n_epochs_scANVI)
+
+    # extract info from posterior
+    scanvi_full = trainer_scanvi.create_posterior(trainer_scanvi.model, net_adata, indices=np.arange(len(net_adata)))
+    latent, _, _ = scanvi_full.sequential().get_latent()
+
+    adata.obsm['X_emb'] = latent
 
+    return adata#, scanvi_full.sequential(), trainer_scanvi
 def runMNN(adata, batch, hvg = None):
```
Comment (LuckyMD) [inline scIB/integration.py:277]: add 2 new lines here

**Review by LuckyMD — CHANGES_REQUESTED** [review]
In general it looks good, but some changes needed: don't scale input data, `use_cuda` default set to True, and a bit of cleanup.

**Comment by LuckyMD** [comment]
> tested and it's working fine

tested in Snakemake or separately?

File: scIB/integration.py (hunk 7 above)
Comment (danielStrobl) [reply scIB/integration.py:195]: that's true for almost every function I think

File: scIB/integration.py (hunk 8 above)
Comment (danielStrobl) [reply scIB/integration.py:230]: That's the value he proposed here: https://github.com/theislab/scib/pull/132#issuecomment-668870534

File: scIB/integration.py (hunk 9 above)
Comment (danielStrobl) [reply scIB/integration.py:251]: I've run it on `icb-sarah`which doesn't have a GPU as far as I know, but yes, setting this to False might be sensible

File: config.yaml [hunk 11]
```diff
@@ -58,6 +58,9 @@ METHODS: # method name : ouput type(s)
   scgen:
     output_type: full
     use_celltype: true
+  scanvi:
+    output_type: embed
+    use_celltype: true
```
[the remark this answers is not in the document]
Comment (danielStrobl) [reply config.yaml:63]: yes, we should deactivate scaling here I think

**Comment by danielStrobl** [comment]
> > tested and it's working fine
> 
> tested in Snakemake or separately?

tested the `runIntegration.py`script

File: scIB/integration.py (hunk 8 above)
Comment (LuckyMD) [reply scIB/integration.py:230]: okay, let's go with it!

File: scIB/integration.py (hunk 7 above)
Comment (LuckyMD) [reply scIB/integration.py:195]: legacy argument i guess... let's remove here though... as no reason to add to the legacy code.

File: scIB/integration.py (hunk 9 above)
Comment (LuckyMD) [reply scIB/integration.py:251]: maybe test briefly via snakemake cpu queue as well.

**Comment by danielStrobl** [comment]
ok, will quickly test on the queue and then it should be ready to merge I guess

**Review by LuckyMD — APPROVED** [review]
looks great now! Confirm queue test results are good and let's merge!

File: scIB/integration.py (hunk 5 above)
Comment (LuckyMD) [reply scIB/integration.py]: I will be careful to mention this in the methods section and if space permits also allude to it in the discussion. Thanks for your help, @romain-lopez!

Comment (romain-lopez) [reply scIB/integration.py]: Great thanks!

**Comment by danielStrobl** [comment]
Tested on snakemake and metrics test run completed now
